Compile TSIC to cvar (fixes #31)

robmck-ms opened the pull request:
This change compiles TSIC to cvar during compile_instructions(). I've verified that it produces an identical cvar to VTT on all our fonts that have a TSIC.

robmck-ms committed:
Code to compile TSIC to cvar

robmck-ms committed:
Add test for TSIC compile

robmck-ms committed:
Make lint happy

robmck-ms committed:
Merge remote-tracking branch 'origin/master' into tsic

madig commented:
Cool! Will check next week. I remember trying to turn TSIC into cvar but failing for some reason. Since VTT compiles it for us, I decided to just dump that.

madig commented:
Hm. What do you think of https://github.com/daltonmaag/vttLib/blob/master/src/vttLib/transfer.py#L21-L24? Have you tested this with other fonts? Not sure if I stumbled over something else there but it dumbfounded me... If we can actually rely on TSIC, we can stop dealing with cvar in dump_to_file and merge_from_file.

robmck-ms commented:
I scratched my head trying to understand that comment. TSIC is a superset of cvar as it contains the final intended values at each varied point in the designspace as well as optional names for designspace points that aren't persisted to cvar. To be sure, I doublechecked with @paullinerud, who wrote the TSIC support in VTT (he and I are on the same team). VTT only builds strictly from TSIC and cvar should contain a subset of regions and CVTs in TSIC. 

Regarding the font in the comment: The only thing I can think of is that either the font brings up a bug in VTT (in which case it would be great to get the font to Paul to debug), or the cvar table was created somewhere else - either hand-made, or pulled in from another font. One interesting thing to try is to round trip it through VTT: delete TSIC, load it in VTT, which will try to generate a TSIC, and then have VTT compile cvar from that. I think it is possible for the cvar to have tuples and triangle functions that are different from what VTT's port of fontTools.models would generate, which might confuse it into generating a TSIC that doesn't match the cvar. 

I tested this PR on the 6 Microsoft VFs that have a TSIC and it produced identical cvars to VTT. Is there another font you'd recommend I test with?

madig commented:
Hmmm!!! I don't have that particular font anymore, but I'm willing to believe that it was a fluke or I messed/mixed something up or something.

Ready to merge then, with one caveat: ~~can you please look into tossing `cvar` (and as a fontTools depdency, `cvt `) dumping and merging in transfer.py and making it `TSIC` only? Otherwise the PR will have to sit here for a while until I can get to it.~~ I need to figure out if I should make `cvar` loading optional, since I think I have at least one dump lying around with it 😣

robmck-ms commented:
I was thinking of the following:

- dumpfile - by default exports both cvar and TSIC (no harm done having both).
- mergefile - by default imports only TSIC, but has --keep-cvar to import cvar optionally.
- compile - by default compiles TSIC, overwriting cvar if present, but has --keep-cvar to ignore TSIC.

Then, you can still use old dumps (with --keep-cvar), but the default workflow focuses on TSIC. 

Alternative: 

- dumpfile - by default exports both cvar and TSIC (same as above).
- mergefile - by default imports both cvar and TSIC.
- compile - by default uses the existing cvar, if present. If cvar is not present, or --compile-TSIC is specified, then it will compile TSIC.

That gives you backward compatibility, and an option for TSIC compilation, but doesn't really nudge users towards using TSIC. Obviously, my preference would be the former.

madig commented:
I also want TSIC to be the source of truth. I'll think about option 1 🤔

robmck-ms commented:
Agreed! That is something I'm working on in our workflows. I want font building to be more like software, where there is a file that is the authoritative source, then an automated pipeline that generates the final ready-to-ship TTF. 

Of course, better yet would be to just be able to run VTT directly to build from the TSI tables (as that's the true source), but that's a whole other project.

madig commented:
(I don't know when I'll get to this PR, if you want to speed up the process, feel free to impl option 1)

robmck-ms commented:
Ok. I'll give it a go. not sure I'll get to it this week.

FYI: in a separate branch, I'm working on SOFFSET/SANCHOR; just need to finish the tests. Also found some other bugs trying to compile [Selawik](https://github.com/unicode-org/text-rendering-tests/blob/master/fonts/Selawik-variable.ttf) (e.g. "JROT[], *, *" doesn't work), but they're further down the work queue.

madig commented:
Cool, thanks! Yeah SOFFSET is the one instruction I only ever found in fonts when something went wrong 😬 I'll take all the fixes you can write up!

robmck-ms committed:
Add --keep-cvar

robmck-ms committed:
Add Noto*_diff_TSIC.ttx which should have been in last commit

robmck-ms committed:
test_TSIC_compile() now compares against a VTT compiled reference

robmck-ms committed:
Make lint happy

robmck-ms commented:
I implemented option 1. I also changed the vttLib_test.test_TSIC_compile() test to compare against a cvar I compiled with VTT to verify compatibility.

madig commented on src/vttLib/__init__.py:
```diff
@@ -706,14 +710,58 @@ def update_composites(font, glyphs=None, vtt_version=6):
         raise VTTLibError("The font contains no 'TSI1' table")
+    if keep_cvar and "cvar" not in font:
+        raise VTTLibError(
+            "The keep_cvar parameter is set, but cvar table exists in the font"
```
```suggestion
            "The keep_cvar parameter is set, but the cvar table is missing from the font"
```

madig commented on src/vttLib/__main__.py:
```diff
@@ -53,10 +53,12 @@ def main(args=None):
     parser_dumpfile.add_argument("outfile", nargs="?", metavar="OUTPUT.ttx")
     parser_dumpfile.set_defaults(func=vttLib.vtt_dump_file)
 
+    parser_mergefile.add_argument("--keep-cvar", action="store_true", default=False)
```
```suggestion
    parser_mergefile.add_argument("--keep-cvar", action="store_true")
```

`store_true` has the side-effect of `default=False`.

Please also add `help="Ignore TSIC and keep the cvar table unchanged."`

madig commented on src/vttLib/transfer.py:
```diff
@@ -60,16 +62,23 @@ def merge_from_file(font: fontTools.ttLib.TTFont, path: os.PathLike) -> None:
 
     TABLES_TO_MERGE = ("TSI0", "TSI1", "TSI2", "TSI3", "TSI5")
-    TABLES_TO_MERGE_OPTIONAL = ("TSIC", "cvt ", "cvar")
+    tables_to_merge_optional = ["TSIC", "cvt "]
```
I think `cvt ` is only needed for `cvar`, because fT adjusts some `cvar` stuff while looking at `cvt `, so try moving `"cvt "` down into the `keep_cvar` guard.

robmck-ms replied:
cvt is still required because all the glyph programs in the glyf table reference it. It's also used when compiling TSIC as the source from which the deltas are made (__init__.py:728-736,746). (cvt is to cvar what glyph outlines are to gvar). There is no other source for this data, except the Control Program from TSI3, but that gets into the realm of being a high-level language compiler instead of just an assembler.

madig replied:
Wait, don't we compile/copy a fresh cvt from TSIwhatever? 🤔 I forgot about that part of the code base.

robmck-ms replied:
I hadn't noticed this code. I've made the change and updated vttLib_test.test_TSIC_compile() to also verify the cvt table against the VTT version. 

This makes the inclusion of the cvt table governed by --keep-cvar. We could similarly add a --keep-cvt to both merge and compile. I'm fine leaving it as is for now. Not sure why I'd want one but not the other...

madig replied:
Thanks! Let's just do the existing cvar switch, cvt was only roped in because of a technicality.

robmck-ms committed:
Update src/vttLib/__init__.py

Co-authored-by: Nikolaus Waxweiler <madigens@gmail.com>

robmck-ms committed:
Update src/vttLib/__main__.py

Co-authored-by: Nikolaus Waxweiler <madigens@gmail.com>

robmck-ms committed:
merge: cvt table also excluded by default

madig commented:
Thanks for working on this 😃
